aip/general/0193.md

```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+Bad examples:
 
-##### Dynamic variables
+- `THE_BOOK_YOU_WANT_IS_NOT_AVAILABLE` (overly verbose)
+- `ERROR` (too general)
 
-The best, actionable error messages include dynamic segments. These
-variable parts of the message are specific to a particular request.
-Without such context, it is unlikely that the message will be fully
-actionable by the user.
+The `domain` field is the logical grouping to which the `reason` belongs.
+The error domain  typically the registered service name of the tool or product
+that  generated the error. The domain must be a globally unique value, and is
+typically the service name, e.g. `pubsub.googleapis.com`.
 
-This practice is critical so that machine actors do not need to rely on
-`LocalizedMessage.message`, which is subject to change and is not part
-of the API contract.
+The (reason, domain) pair form a machine-readable way of identifying a particular error.
+Services **must** use the same (reason, domain) pair for the same error, and
+**must not** use the same (reason, domain) pair for logically different errors.
+The decision about whether two errors are "the same" or not is not always clear, but
+should generally be considered in terms of the expected action a client might take
+to resolve them.
 
-Consider the following example:
+The `metadata` field is a map of key/value pairs providing additional
+dynamic information as context. Each key within `metadata` **must** be at most
+64 characters long, and conform to the regular expression `[a-z][a-zA-Z0-9-_]+`.
 
-> The Book, "The Great Gatsby", is unavailable at the Library, "Garfield
-> East".  It is expected to be available again on 2199-05-13.
+Any request-specific information which contributes to the `Status.message` or
+`LocalizedMessage.message` messages **must** be represented within `metadata`.
+This practice is critical so that machine actors do not need to parse error
+messages to extract information.
 
-The preceding error message is made actionable by the context, both
-originating from the request, the title of the Book, the name of the
-Library, and by the information that is known only by the service, that
-is, the expected return date of the Book.
+For example consider the following message:
 
-All dynamic variables found in error messages **must** also be present
-in the `map<string, string>`, `ErrorInfo.metadata` (found on the
-*required* `ErrorInfo`). For example, the `metadata` map for the sample
-error message above will include *at least* the following key/value
-pairs:
+> An &lt;e2-medium&gt; VM instance with &lt;local-ssd=3,nvidia-t4=2&gt; is currently unavailable
+> in the &lt;us-east1-a&gt; zone. Consider trying your request in the &lt;us-central1-f,us-central1-c&gt;
+> zone(s), which currently has/have capacity to accommodate your request. Alternatively,
+> you can try your request again with a different VM hardware configuration
+> or at a later time. For more information, see the troubleshooting documentation.
```

Comment:
I think there was some copy paste format issues here with the `&lt;e2-medium&gt;` stuff

Reply:
Nope, this was deliberate. That's how I'd expect to represent `<` and `>` in Markdown when they're not in code. I don't know what the AIP site generator will do, but it's deliberately as it is :)

Reply:
TIL! let's see what happens :)

Reply:
I suggest checking how these HTML character render. I'm showing a blank for the rendered zone name in the following text:

Consider trying your request in the , zone(s), which currently has/have capacity to accommodate your request. Alternatively, you can try your request again with a different VM hardware configuration or at a later time.

Reply:
Basically I don't know what the site generator will do - it's possible that backslashes are the way forward.

@noahdietz: How easy is it run the site generator, and will it involve installing a load of things? (If you have it set up locally and could test this, that would be really handy.)

Reply:
@jskeet there are instructions on how to run it locally here: https://google.aip.dev/contributing#local-installation

Though I haven't been able to get it running myself - not sure if it is a python version thing or not. It's a black box to me tbh

Reply:
I've actually got it working now - and it looks like using the HTML escaping works, whereas backslashes don't. Once we've merged, we should immediately double check though...